src/daemon/socks5.rs

```diff
@@ -0,0 +1,121 @@
+use std::{
+    net::{Ipv4Addr, SocketAddr, SocketAddrV4, ToSocketAddrs},
+    str::FromStr,
+};
+
+use anyhow::Context;
+use earendil_crypt::Fingerprint;
+use futures_util::io;
+use smol::{
+    future::FutureExt,
+    io::AsyncWriteExt,
+    net::{TcpListener, TcpStream},
+};
+use smolscale::reaper::TaskReaper;
+use socksv5::v5::*;
+
+use crate::{
+    config::{Fallback, Socks5},
+    socket::{Endpoint, Socket},
+    stream::Stream,
+};
+
+use super::DaemonContext;
+
+pub async fn socks5_loop(ctx: DaemonContext, socks5_cfg: Socks5) -> anyhow::Result<()> {
+    log::debug!("socks5 loop");
+    let tcp_listener = TcpListener::bind(SocketAddrV4::new(
+        "127.0.0.1".parse()?,
+        socks5_cfg.listen_port,
+    ))
+    .await?;
+    let fallback = socks5_cfg.fallback;
+    let reaper = TaskReaper::new();
+
+    loop {
+        let ctx = ctx.clone();
+        let fallback = fallback.clone();
+        let (tcp_stream, _) = tcp_listener.accept().await?;
+
+        reaper.attach(smol::spawn(async move {
+            tcp_stream.set_nodelay(true)?;
```

Comment:
the naming here is a bit confusing --- instead of `tcp_stream` and `proxy_stream` something like `client` and `remote` seem better